File: wingc/test-files/composite-simple.py
from wing_rpc import Schema, Enum
from typing import ClassVar
from enum import StrEnum


class Result(Schema):
    __match_args__: ClassVar[tuple] = ('id', 'title', 'description',)
    id: int
    title: str
    description: str


class ByName(Schema):
    __match_args__: ClassVar[tuple] = ('name',)
    name: str


class ById(Schema):
    __match_args__: ClassVar[tuple] = ('id',)
    id: int


class Search(Enum):
    __match_args__: ClassVar[tuple] = ('tag', 'value',)
    class Tag(StrEnum):
        ByName = 'ByName'
        ById = 'ById'
    tag: Tag
    value: ByName | ById


class Message(Enum):
    __match_args__: ClassVar[tuple] = ('tag', 'value',)
    class Tag(StrEnum):
        Search = 'Search'
        results = 'results'
    tag: Tag
    value: ById | list[Result]

File: wingc/test-files/composite-simple.rs
use serde::{Serialize, Deserialize};
use wing_rpc::Message as WingMessage;


#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Result {
    pub id: usize,
    pub title: String,
    pub description: String,
}

impl<'a> WingMessage<'a> for Result {
    const NAME: &'static str = "Result";
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ByName {
    pub name: String,
}

impl<'a> WingMessage<'a> for ByName {
    const NAME: &'static str = "ByName";
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ById {
    pub id: usize,
}

impl<'a> WingMessage<'a> for ById {
    const NAME: &'static str = "ById";
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Search {
    ByName(ByName),
    ById(ById),
}

impl<'a> WingMessage<'a> for Search {
    const NAME: &'static str = "Search";
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Message {
    Search(ById),
    results(Vec<Result>),
}

impl<'a> WingMessage<'a> for Message {
    const NAME: &'static str = "Message";
}

